### ISSUE TITLE
Pass xarray to napari-gui for autolabeling sliders

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#14]
xarray is a library for N-D labeled arrays and datasets in Python http://xarray.pydata.org. If we allow xarrays to be passed to napari-gui then we could autopopulate labels of sliders etc. based on the names of the dimensions in the xarray object

**Comment by jni** [comment]
Just to clarify, I think xarray support would be brilliant, but I just want to make sure that we don't become *dependent* on it. Bare NumPy arrays should be first-class citizens.

**Comment by AhmetCanSolak** [comment]
@jni can you share why you think in that way? Just curious.

**Comment by kne42** [comment]
This sounds like a nice convenience functionality; I'll keep it in mind!

Also, I'd like to note that you can already use `DataArray`s and other `numpy.ndarray`-like objects since we never explicitly call any NumPy functions on them (which would autoconvert them to strict `ndarray`s).

**Comment by jni** [comment]
@AhmetCanSolak the entire SciPy ecosystem has thrived by using NumPy arrays as a lingua franca between packages. We are currently seeing some fragmentation as people test the boundaries of what can be done with arrays. I don't want to contribute to this fragmentation, and NumPy itself is taking steps to address it (see [NEP18](https://www.numpy.org/neps/nep-0018-array-function-protocol.html)). There is a good chance that supporting NumPy arrays will eventually let us support any array in the future "for free". The converse is not true.

**Comment by AhmetCanSolak** [comment]
Thank you @jni !
It is nice to see people are considering this in NEP18. 

For our use, I dont think we will have issues with either labeled or unlabeled arrays, as soon as we document it well and be strict on our implementations.

**Comment by ambrosejcarr** [comment]
> Just to clarify, I think xarray support would be brilliant, but I just want to make sure that we don't become dependent on it. Bare NumPy arrays should be first-class citizens.

I agree here. Numpy should always be first class. As a napari user, I want labeled axes in Napari so I know what axis of my 5-d tensor I am toggling (these tensors _happen_ to be in xarray). I would suggest allowing napari to take an iterable of string labels whose order match the sliders produced by showing an image with dimensionality > 2. 

Later, xarray support could be added such that when an xarray is passed, it auto-populates these labels. If you point me at the relevant place in the codebase, I would be happy to make these contributions. It's a pain point in our workflows (see: https://github.com/spacetx/starfish/blob/master/starfish/display/stack.py#L194).

**Comment by HagaiHargil** [comment]
Hi,

I would also like to see support for labeled dimensions and automatically detecting `xarray`'s given labels. I'll be happy to write the PR for this feature, but I will need a bit of mentoring. I'm assuming we'd want to add a `label` attribute to the range sliders defined [here](https://github.com/napari/napari/blob/4499372a6256fc6d2cd4b3051a55297e258368ff/napari/_qt/qt_range_slider.py)? And if a `label` is not `None`, rescale the drawn slider to accommodate for the given label?

**Comment by sofroniewn** [comment]
Hi @HagaiHargil, great that you're excited about this.

I think it will be best to split this into two PRs, the first PR will just add support for labeled axes if someone passes in a list of string labels, and the second PR will add support for autodetecting xarrays labels and generating that string list. How does that sound?

The reason for this is that we'll always want to support people with just numpy arrays, but want to provide them access to this functionality without having to make an xarray.

For the first PR - here's some initial guidance on how I'm thinking about it, happy to provide more as you go:

Let's leave the `QLabel` off of the range slider explicitly and instead we'll add it to the layout in `qt_dims.py`, around the same place as we add the dimensions slider https://github.com/napari/napari/blob/b8b02a4642acad973e67228acc55e7cabef85346/napari/_qt/qt_dims.py#L229 (which is actually a QScrollBar).

We'll also want to add the labels as a list of strings to our dimensions model in `dims.py`, similar to say the points property here https://github.com/napari/napari/blob/4499372a6256fc6d2cd4b3051a55297e258368ff/napari/components/dims.py#L95., with the appropriate setters and events like those other properties. It can default to be a list of empty strings.

The dimensions model constructor should also take an optional keyword argument `axis_labels=None` that if provided is a list the same length as `init_ndim`.

The `qt_dims` view should then be updated with events when the labels change to update the display of the label in the bottom left next to the qscrollbar area. We're probably also going to be adding a play button there when #607 is done, which might make this PR a bit easier on the qt layout side.

I think probably best then to also expose the `axis_labels` at the `Viewer` level, at which point you'd need to add it to `Viewer.py`, `ViewerModel.py` and all our `view_*` methods. Inside the `ViewerModel` you might need to add some checks around the `ndim` calculation and then pass it to the Dims model - here https://github.com/napari/napari/blob/4499372a6256fc6d2cd4b3051a55297e258368ff/napari/components/viewer_model.py#L65. Looking at that code we should also add `ndisplay=2` and `order=None` to the Dims model keywords args while we're at it too.

Once we get this PR merged we can then work on the best way to autodetect the xarray labels and propagate them from the layers to the viewer.

Let me know how this all sounds.

**Comment by HagaiHargil** [comment]
Hey, and thanks for this detailed reply.

Splitting up the PRs seems very logical. I will start working on the first soon, and I might create a semi-complete PR for you to review, and then update it with your comments.

**Comment by ianhi** [comment]
Hi,

Whats the current status of this? I see that named sliders are now supported, so I think we would now need to the second PR.

I'd like to help with this, but it's not clear to me if this would need to wait until after issues such as https://github.com/napari/napari/issues/906 are resolved. 

----

Also, it would awesome to also label the slider readouts in addition to the slider names.  @jrussell25 implemented this for ipywidgets/matplotlib sliders into [hyperslicer](https://mpl-interactions.readthedocs.io/en/stable/examples/hyperslicer.html#Hyperslicer-with-Xarray) which autopopulates sliders with names based on the dims and if there are coords it uses that to fill out slider label. So 
```python
%matplotlib ipympl
import matplotlib.pyplot as plt
import numpy as np
import xarray as xr
from mpl_interactions import hyperslicer

rng = np.random.default_rng()
data = xr.DataArray(
    rng.random((3, 5, 512, 512)),
    dims=("channel","z", "x", "y"),
    coords={"z":np.linspace(-1.5, 1.5, 5), "channel": ["BF", "GFP", "mCherry"]},
)

plt.figure()
ctrls = hyperslicer(data)
```

gives this (note the slider readouts):
![Peek 2021-02-07 13-07](https://user-images.githubusercontent.com/10111092/107155308-a8c3db80-6945-11eb-8d60-49b2226f3e3d.gif)


fwiw we accomplish this without ever importing xarray:
https://github.com/ianhi/mpl-interactions/blob/5d6f063a5a1e2a357991bd98e77bd9e80ea5dc81/mpl_interactions/generic.py#L569-L570

https://github.com/ianhi/mpl-interactions/blob/5d6f063a5a1e2a357991bd98e77bd9e80ea5dc81/mpl_interactions/xarray_helpers.py


I'm not sure if napari supports this, but the key step for this is to allow sliders to have a displayed value but also keep track of their index in the array. For ipywidgets we accomplished this by squashing a label and slider together and then linking them with a function that transforms the slider value (an integer index) to the value of the coords.
https://github.com/ianhi/mpl-interactions/blob/5d6f063a5a1e2a357991bd98e77bd9e80ea5dc81/mpl_interactions/helpers.py#L363-L369 which maybe would be possible with qt as well?

**Comment by sofroniewn** [comment]
> Whats the current status of this? I see that named sliders are now supported, so I think we would now need to the second PR.
>
> I'd like to help with this, but it's not clear to me if this would need to wait until after issues such as #906 are resolved.

I think if we can resolve first #906 then implement this will be easier. Maybe we continue a discussion on #906 now, there's nothing blocking that - I think we just need to think about exactly how we want everything to work. If you were willing to drive this @ianhi I'm sure we'd all be supportive.

> Also, it would awesome to also label the slider readouts in addition to the slider names. @jrussell25 implemented this for ipywidgets/matplotlib sliders into hyperslicer which autopopulates sliders with names based on the dims and if there are coords it uses that to fill out slider label.

We have some ability to do that for things like our contrast_limits slider (if you right click then you get a pop-up that can be edited further) and our dims sliders thanks to @tlambert03 (see the little edit boxes with numbers you can type into next to the sliders) but it is not standard for all our sliders (like opacity say)

![image](https://user-images.githubusercontent.com/6531703/107157296-28e34480-6938-11eb-8117-d6900c9b7106.png)

We have #879 that wants to bring this too all sliders (though I was recently discussing with @goanpeca and @ppwadhwa that we might not want to rely on pop-ups, but instead just have the value always displayed. I wonder also @tlambert03 if this "slider with value" might one day be a special magicgui widget type.

**Comment by ianhi** [comment]
> We have some ability to do that for things like our contrast_limits slider (if you right click then you get a pop-up that can be edited further) and our dims sliders thanks to @tlambert03 (see the little edit boxes with numbers you can type into next to the sliders) but it is not standard for all our sliders (like opacity say)

I may be wrong but I think that that is a little bit different than what I am proposing. I'd like to be able to do something like this:
`viewer.add_image(3d_array, axis_labels=["time", "y", "x"], axis_values = {"time": np.arange(0, 60, 15)})`

So that the value of the slider used in callbacks to update the display will still be an integer used for indexing, but the value displayed to the user is taken from `axis_values`.

What I'm proposing would look like the third slider in this GIF:
![Peek 2021-02-07 15-21](https://user-images.githubusercontent.com/10111092/107158455-27c20f80-6958-11eb-9be2-4f815d91b25e.gif)


I *think* that is a bit different than the pop-up you're talking about, but I could definitely be misinterpreting. If it is different then I don't think that this will be totally covered by #906

**Comment by tlambert03** [comment]
yeah, we don't have that sort of transform from "slider space" to "user-defined-unit-space" yet.

**Comment by sofroniewn** [comment]
> I think that is a bit different than the pop-up you're talking about, but I could definitely be misinterpreting.

Yup sorry, I totally missed the `coords={"z":np.linspace(-1.5, 1.5, 5)}` part. We don't have anything like that yet and it's not really captured well in any of the current issues. Maybe @ianhi you want to make a dedicated issue to discuss just that functionality

**Comment by ianhi** [comment]
> not really captured well in any of the current issues. Maybe @ianhi you want to make a dedicated issue to discuss just that functionality

Would this be better housed on napari or on magicgui?

**Comment by tlambert03** [comment]
I think napari is still the better place for this.  We don't use magicgui for our dims sliders, so we'd need to change a lot before that saw the light of day.

**Comment by manzt** [comment]
> Yup sorry, I totally missed the coords={"z":np.linspace(-1.5, 1.5, 5)} part. We don't have anything like that yet.

I would _love_ something like this. I have a handful of heatmaps that overlay a WSI, and the UI isn't great to toggle each of these as independent 2-D layers. The other option is to stack the heatmaps as single array`shape=(n_heatmaps, h, w)` and use the axis slider to flip between heatmaps. The only issue is that I lose the ability to communicate the labels for each heatmap in the UI. (in this example, the `coords` for the heatmap dimension are categorical labels for the modality corresponding to the heatmap.)

**Comment by ianhi** [comment]
>  Maybe @ianhi you want to make a dedicated issue to discuss just that functionality

https://github.com/napari/napari/issues/2246

@manzt if all you need to is view the layers you should try out [hyperslicer](https://mpl-interactions.readthedocs.io/en/stable/examples/hyperslicer.html#Hyperslicer-with-Xarray). The downside is that it's based in matplotlib so it's lower performance, and you can't do any of the fancy things napari does. But for quick viewing it is still helpful (not obvious from the docs but it will work both in and out of jupyter notebooks). 

My plan is to keep using both napari and hyperslicer until this lands in napari and then I will only use napari.

**Comment by manzt** [comment]
Thanks for the suggestion! I'll check it out. I think the main issue is that heatmaps are scaled and translated to overlay a multiscale RGB image (which is important for interpretation), so I'm pretty dependent on napari features.

**Comment by TomNicholas** [comment]
Chiming in 4 years after the original opening to give the current situation (from an xarray dev's biased perspective). :sweat_smile: (@tlambert03 sent me here from [here](https://github.com/napari/napari/pull/3715#issuecomment-1189218372))

> There is a good chance that supporting NumPy arrays will eventually let us support any array in the future "for free". The converse is not true.

This used to be the case, but now Xarray [explicitly aims](https://docs.xarray.dev/en/stable/roadmap.html#flexible-arrays) to wrap any array type that exposes NumPy's API. To this end we have pushed hard for standardisation in the wider array community  - see for example [all (1)](https://numpy.org/neps/nep-0030-duck-array-protocol.html) [the (2)](https://numpy.org/neps/nep-0037-array-module.html) [NEP's (3)](https://numpy.org/neps/nep-0047-array-api-standard.html) co-authored by @shoyer.

Xarray can currently wrap numpy, dask, pint, sparse, cupy arrays (and probably more that we haven't tried). Thanks to standardisation efforts in the wider community, we should [very soon](https://github.com/pydata/xarray/pull/6804) be able to wrap *any array* that conforms to the new [python array API standard](https://data-apis.org/array-api/latest/index.html).

**Comment by sofroniewn** [comment]
> Xarray can currently wrap numpy, dask, pint, sparse, cupy arrays (and probably more that we haven't tried). Thanks to standardisation efforts in the wider community, we should https://github.com/pydata/xarray/pull/6804 be able to wrap any array that conforms to the new [python array API standard](https://data-apis.org/array-api/latest/index.html).

This is very exciting! As the OP 4-years ago, I'm very happy to hear things are moving in this direction!! Maybe @andy-sweet wants to look into the above and your comment here https://github.com/napari/napari/pull/3715#issuecomment-1189208297 and think about any of this in the context of the work to make slicing async in napari. Thanks!!

**Comment by kevinyamauchi** [comment]
Thanks for the follow-up, @TomNicholas . It was nice meeting you at scipy! Super exciting stuff. I look forward to digging into all of the new developments (especially the flexible indexes).

**Comment by sofroniewn** [comment]
Another good reason to consider better xarray support might be [ndpyramid](https://github.com/carbonplan/ndpyramid) (h/t @GenevieveBuckley who pointed it out in https://github.com/dask/dask-blog/issues/141#issuecomment-1197547317

**Comment by TomNicholas** [comment]
Yes Kevin and I also discussed support for multi-resolution data in Napari.
Ndpyramid wraps xarray-DataTree, a prototype feature addition to xarray. I
am the main person developing xarray-DataTree, so feel free to ping me with
questions about that too!

https://github.com/xarray-contrib/datatree

**Comment by GenevieveBuckley** [comment]
Here is another strong reason for supporting this as a feature: https://twitter.com/jnuneziglesias/status/1564858125772935169

**Comment by psobolewskiPhD** [comment]
I'm going to bump this.
At SciPy 2024 I saw a number of talks showcasing xarray and spoke with @TomNicholas 
Across coordinates, units, various data reductions, lazy loading, and DataTree, xarray has a lot to offer and would not in any way box out or diminish using numpy arrays
I've used xarray via aicsimageio and it's a joy to work with, especially if you dont exactly remember what was what in your data, because it's year(s) after you did the imaging 👀 🤣

**Comment by TomNicholas** [comment]
Thanks @psobolewskiPhD ! 

> Across coordinates, units, various data reductions, lazy loading, and DataTree, xarray has a lot to offer and would not in any way box out or diminish using numpy arrays

I agree 😛 

FOr anyone interested to work on this Earthmover won some [funding](https://figshare.com/articles/preprint/Supporting_the_adoption_of_Xarray_in_the_biomedical_research_community_Proposal_to_the_Chan_Zuckerberg_Initiative_Essential_Open_Source_Software_for_Science_Program/26121199/1?file=47285551) to hire someone to work on exactly this kind of thing (better integrating xarray with biomedical projects). cc @dcherian

**Comment by imagejan** [comment]
What about xarray support for plate-like datasets with rows and columns? We have `xarray.DataArray`s with dimensions/coordinates `row` and `column` (or for biological experiments you could have e.g. `genotype` and `treatment`).

It would be nice to have the dimension sliders labelled accordingly upon adding the array, but also to support viewing a grid over any one or two dimensions.

What I'd love to see is a grid view implementation that takes any two dimensions `dim0` and `dim1` of the current layer(s) and displays a grid of `len(dim0) x len(dim1)` tiles (with all other dimension sliders left untouched). As far as I understand, the current [grid](https://napari.org/dev/tutorials/fundamentals/viewer.html#grid-button) mode separates layers and uses "Grid width" and "Grid height" as parameters, but we could also work on array dimensions instead of layers. Is anybody doing this already?

**Comment by jni** [comment]
> also to support viewing a grid over any one or two dimensions.

Oooh, clever!

> Is anybody doing this already?

Not to my knowledge, and I don't think it would be particularly easy to implement, either. But it's a good idea to keep in mind as we improve our support for xarray and other arrays with labeled axes (which *is* in the works).
